### PR TITLE
Expand Lean Default Charts and Reduce Clutter

### LINKED ISSUE #5612: Lean Default Charts
#### Expected Behavior

- Lean default charts always provide valuable information
#### Actual Behavior

- Framework relate charts: Alpha, Insight Count, Alpha Assets are always empty for non framework algorithms
#### Potential Solution

- Only chart/store the mentioned framework related charts when they have information, at least 1 point != default value
- Add new charts:
        - Drawdown underwater plot [already present in Report generator]
        - Long & Short ratio [already present in Report generator]
        - `Alpha Assets` -> `Assets`. Can be alpha/framework or trade/non framework based
        - Strategy capacity
#### Reproducing the Problem

N/A
#### System Information

N/A
#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by C-SELLERS**
#### Description

- Refactor ChartingInsightManagerExtension.cs  
    - Stop Alpha framework from adding charts to algorithm until insight occurs (Fixes empty charts for non-framework based algos)
    - Cleanup, nit fixes, and simplify charting and series handling
- Add new defaults charts: 
    - Drawdown
    - Long Short Exposure
    -  Assets Sales Volume HeatMap
    - Strategy Capacity (Only implemented for BacktestingResultHandler)
- Add some comments for future clarity

#### Related Issue

Closes #5612 

#### Motivation and Context

To make default charts more useful and reduce clutter by stopping alpha charts when they are not needed.

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

All regressions passing
Tested in the cloud, see examples in comments below.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by C-SELLERS** [comment]
For all algorithms referenced below I pushed the end date out 1 year so we could have a nice picture of charting behavior.

BasicTemplateFuturesFrameworkAlgorithm 
image


BasicTemplateOptionsAlgorithm
image


BasicTemplateOptionsFilterUniverseAlgorithm
image


G10CurrencySelectionModelFrameworkAlgorithm 
image

File: Engine/Alphas/ChartingInsightManagerExtension.cs [hunk 1]
```diff
@@ -32,20 +32,24 @@ public class ChartingInsightManagerExtension : IInsightManagerExtension
...
+        private bool _chartsAdded;
+        private IAlgorithm _algorithm;
 
-        private int _dailyCount;
+        // Heatmap chart
+        private readonly Chart _totalInsightCountPerSymbolChart = new Chart(AlphaAssets);          
         private readonly Dictionary<Symbol, int> _totalInsightCountPerSymbol = new Dictionary<Symbol, int>();
-        private readonly Dictionary<InsightScoreType, Series> _insightScoreSeriesByScoreType = new Dictionary<InsightScoreType, Series>();
+
+        private readonly Chart _totalInsightCountChart = new Chart("Insight Count");
```
Comment (Martin-Molinero) [inline Engine/Alphas/ChartingInsightManagerExtension.cs:51]: Unless required suggest avoiding changing the name of existing charts, can cause issues with any existing consumers

File: Engine/Alphas/ChartingInsightManagerExtension.cs [hunk 2]
```diff
@@ -88,19 +86,31 @@ public ChartingInsightManagerExtension(IAlgorithm algorithm, StatisticsInsightMa
...
+            }
+
             // sample insight/symbol counts each utc day change
             if (frontierTimeUtc.Date > _lastInsightCountSampleDateUtc)
             {
                 _lastInsightCountSampleDateUtc = frontierTimeUtc.Date;
 
                 // add sum of daily insight counts to the total insight count series
-                _totalInsightCountSeries.AddPoint(frontierTimeUtc.Date, _dailyCount);
+                _totalInsightCountChart.Series["Count"].AddPoint(frontierTimeUtc.Date, _dailyInsightCount);
```
Comment (Martin-Molinero) [inline Engine/Alphas/ChartingInsightManagerExtension.cs:107]: `_totalInsightCountSeries` was a direct access vs `.Series["Count"].` which causes a lookup on each call?

File: Engine/Results/BacktestingResultHandler.cs [hunk 3]
```diff
@@ -509,6 +514,30 @@ public virtual void RuntimeError(string message, string stacktrace = "")
...
+            var dayChanged = PreviousUtcSampleTime.Date != time.Date;
+            if (dayChanged || force)
+            {
+                // Use previous time for sample unless forced
+                var sampleTime = force ? time : PreviousUtcSampleTime;
+
+                // Sample strategy capacity, round to 1k
+                var roundedCapacity = Math.Round(_capacityEstimate.Capacity.Value / 1000) * 1000;
+                Sample("Capacity", "Strategy Capacity", 0, SeriesType.Line, sampleTime,
+                    roundedCapacity, AlgorithmCurrencySymbol);
```
Comment (Martin-Molinero) [inline Engine/Results/BacktestingResultHandler.cs:534]: Why special behavior for Capacity? maybe if just backtesting we can have SampleCapacity which is `nop` in live trading?

File: Engine/Results/BaseResultsHandler.cs [hunk 4]
```diff
@@ -479,9 +496,7 @@ public virtual void Sample(DateTime time, bool force = false)
         /// <param name="value">Current equity value.</param>
         protected virtual void SampleEquity(DateTime time, decimal value)
         {
-            var accountCurrencySymbol = Currencies.GetCurrencySymbol(Algorithm.AccountCurrency);
-
-            Sample("Strategy Equity", "Equity", 0, SeriesType.Candle, time, value, accountCurrencySymbol);
+            Sample("Strategy Equity", "Equity", 0, SeriesType.Candle, time, value, AlgorithmCurrencySymbol);
```
Comment (Martin-Molinero) [inline Engine/Results/BaseResultsHandler.cs:499]: Love the `AlgorithmCurrencySymbol` caching ✨

File: Engine/Results/BaseResultsHandler.cs [hunk 5]
```diff
@@ -509,6 +524,90 @@ protected virtual void SampleBenchmark(DateTime time, decimal value)
...
+        }
+
+        /// <summary>
+        /// Sample assets sales volume
+        /// </summary>
+        /// <param name="time">Time of the sample</param>
+        protected virtual void SampleSalesVolume(DateTime time)
+        {
+            var volumeBySymbol = Algorithm.Portfolio.Values
+                .Select(x => new KeyValuePair<Symbol, decimal>(x.Symbol, x.TotalSaleVolume));
```
Comment (Martin-Molinero) [inline Engine/Results/BaseResultsHandler.cs:550]: nit - We don't really need the select operation here?

Think we should probably skip if `TotalSaleVolume` is 0, thinking algorithms selecting many symbols or custom symbols

File: Engine/Results/BaseResultsHandler.cs [hunk 6]
```diff
@@ -509,6 +524,90 @@ protected virtual void SampleBenchmark(DateTime time, decimal value)
...
+                    longHolding.Value/currentPortfolioValue, "");
+            }
+        }
+
+        /// <summary>
+        /// Helper method for SampleExposure which separates security holdings by type and then sums their holdings
+        /// </summary>
+        /// <param name="holdings"></param>
+        /// <returns></returns>
+        private static Dictionary<SecurityType, decimal> SumHoldingsBySecurityType(IEnumerable<SecurityHolding> holdings)
```
Comment (Martin-Molinero) [inline Engine/Results/BaseResultsHandler.cs:590]: nit - Maybe can replace `SumHoldingsBySecurityType` for a call to `Aggregate`?

File: Engine/Results/BaseResultsHandler.cs [hunk 7]
```diff
@@ -509,6 +524,90 @@ protected virtual void SampleBenchmark(DateTime time, decimal value)
...
+            {
+                Sample("Exposure", $"{shortHolding.Key} - Short", 0, SeriesType.Line, time,
+                    shortHolding.Value/currentPortfolioValue, "");
+            }
+
+            var longBySecurityType = SumHoldingsBySecurityType(longHoldings);
+            foreach (var longHolding in longBySecurityType)
+            {
+                Sample("Exposure", $"{longHolding.Key} - Long", 0, SeriesType.Line, time,
+                    longHolding.Value/currentPortfolioValue, "");
```
Comment (Martin-Molinero) [inline Engine/Results/BaseResultsHandler.cs:581]: Might want to check `currentPortfolioValue` isn't 0 👍 💥

**Review by Martin-Molinero — COMMENTED** [review]
Sweet improvement, leaving a few minor requests

File: Engine/Alphas/ChartingInsightManagerExtension.cs (hunk 2 above)
Comment (C-SELLERS) [reply Engine/Alphas/ChartingInsightManagerExtension.cs:107]: Thats true, I'll switch back and keep the series as a class var 👍🏽

File: Engine/Alphas/ChartingInsightManagerExtension.cs (hunk 1 above)
Comment (C-SELLERS) [reply Engine/Alphas/ChartingInsightManagerExtension.cs:51]: All the names are the same! Just got a moved around a bit I promise 😁

File: Engine/Results/BaseResultsHandler.cs (hunk 6 above)
Comment (C-SELLERS) [reply Engine/Results/BaseResultsHandler.cs]: Opt'd for group by then sum, not sure how to apply aggregate for this case 😁

**Comment by C-SELLERS** [comment]
Push all the end dates out one year for these tests

Tested on Lean v12040

BasicTemplateCryptoAlgorithm
image


BasicTemplateFuturesFrameworkAlgorithm
image


BasicTemplateOptionsAlgorithm
image

BasicTemplateOptionsFilterUniverseAlgorithm
image

DropboxUniverseSelectionAlgorithm
image

FuturesMomentumAlgorithm
image

G10CurrencySelectionModelFrameworkAlgorithm
image

MACDTrendAlgorithm
image

File: Engine/Results/BaseResultsHandler.cs [hunk 8]
```diff
@@ -509,6 +525,89 @@ protected virtual void SampleBenchmark(DateTime time, decimal value)
...
+
+        /// <summary>
+        /// Sample assets sales volume
+        /// </summary>
+        /// <param name="time">Time of the sample</param>
+        protected virtual void SampleSalesVolume(DateTime time)
+        {
+            foreach (var holding in Algorithm.Portfolio.Values)
+            {
+                if (holding.TotalSaleVolume != 0)
```
Comment (Martin-Molinero) [inline Engine/Results/BaseResultsHandler.cs:552]: `if (holding.TotalSaleVolume != 0)` + We might want to order by TotalSaleVolume and cap the quantity being sampled, thinking about an algorithm trading multiple equities options, securities will grow quite big in time, say QC500 + options

File: Engine/Results/BaseResultsHandler.cs [hunk 9]
```diff
@@ -509,6 +525,89 @@ protected virtual void SampleBenchmark(DateTime time, decimal value)
...
+        /// <param name="type">Side to sample from portfolio</param>
+        /// <param name="time">Time of the sample</param>
+        /// <param name="currentPortfolioValue">Current value of the portfolio</param>
+        private void SampleExposureHelper(PositionSide type, DateTime time, decimal currentPortfolioValue)
+        {
+            // Shorts = holdings < 0; Longs = holdings > 0
+            var multiplier = type == PositionSide.Long ? 1 : -1;
+
+            // Select the holdings that fit our position side
+            var positionsBySecurityType = Algorithm.Portfolio.Values.Where(holding => multiplier * holding.HoldingsValue > 0)
```
Comment (Martin-Molinero) [inline Engine/Results/BaseResultsHandler.cs:591]: nit - `holding => multiplier * holding.HoldingsValue > 0` could directly use Invested/HoldStock properties?

File: Engine/Results/BacktestingResultHandler.cs [hunk 10]
```diff
@@ -572,6 +577,18 @@ protected override void SampleEquity(DateTime time, decimal value)
...
         }
 
+        /// <summary>
+        /// Sample estimated strategy capacity
+        /// </summary>
+        /// <param name="time">Time of the sample</param>
+        protected override void SampleCapacity(DateTime time)
+        {
+            // Sample strategy capacity, round to 1k
+            var roundedCapacity = Math.Round(_capacityEstimate.Capacity.Value / 1000) * 1000;
```
Comment (Martin-Molinero) [inline Engine/Results/BacktestingResultHandler.cs:587]: nit - think extension method `DiscretelyRoundBy` does something similar?

File: Engine/Results/BaseResultsHandler.cs [hunk 11]
```diff
@@ -547,13 +547,12 @@ protected virtual void SampleDrawdown(DateTime time, decimal currentPortfolioVal
...
+            foreach (var holding in Algorithm.Portfolio.Values.Where(y => y.TotalSaleVolume != 0)
+                .OrderByDescending(x => x.TotalSaleVolume).Take(30))
             {
-                if (holding.TotalSaleVolume != 0)
-                {
-                    Sample("Assets Sales Volume", $"{holding.Symbol.Value}", 0, SeriesType.Treemap, time,
-                        holding.TotalSaleVolume, AlgorithmCurrencySymbol);
-                }
+                Sample("Assets Sales Volume", $"{holding.Symbol.Value}", 0, SeriesType.Treemap, time,
+                    holding.TotalSaleVolume, AlgorithmCurrencySymbol);
```
Comment (Martin-Molinero) [inline Engine/Results/BaseResultsHandler.cs:556]: Cool! 👍

**Review by Martin-Molinero — APPROVED** [review]

**Review by C-SELLERS — COMMENTED** [review]
Just did a final self-review, 👍🏽

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**